File: src/lib/progression.test.ts
import { describe, it, expect } from "vitest";
import {
  buildAllProgressions,
  buildExerciseProgression,
  estimateOneRepMax,
  loadForReps,
  loadIncrementLbs,
  parseRepRange,
  parseSetTarget,
  prescribeNextSession,
  prescribeWorkoutDay,
  roundToLoadable,
  summarizeProgressions,
} from "./progression";
import type { WorkoutExercise, WorkoutSetLog } from "./types";

function log(
  date: string,
  exerciseName: string,
  setIndex: number,
  weightLbs?: number,
  reps?: number,
  rpe?: number,
  section: WorkoutSetLog["section"] = "main",
): WorkoutSetLog {
  return {
    id: `${date}:${exerciseName}:${setIndex}`,
    date,
    planId: "plan-1",
    dayLabel: "Monday",
    section,
    exerciseName,
    globalSlot: 0,
    setIndex,
    weightLbs,
    reps,
    rpe,
    loggedAt: `${date}T18:00:00.000Z`,
  };
}

const benchPress: WorkoutExercise = { name: "Bench Press", sets: "3", reps: "8-12" };

describe("estimateOneRepMax", () => {
  it("uses Epley when no RPE is given", () => {
    // 200 x 5 → 200 * (1 + 5/30) = 233.3
    expect(estimateOneRepMax(200, 5)).toBeCloseTo(233.33, 1);
  });

  it("credits reps in reserve when RPE is logged", () => {
    // 5 @ RPE 8 has ~2 in reserve → treated as 7 effective reps
    expect(estimateOneRepMax(200, 5, 8)).toBeCloseTo(246.67, 1);
    // RPE 10 = taken to failure = plain Epley
    expect(estimateOneRepMax(200, 5, 10)).toBeCloseTo(estimateOneRepMax(200, 5), 5);
  });

  it("returns 0 for missing or invalid input", () => {
    expect(estimateOneRepMax(0, 5)).toBe(0);
    expect(estimateOneRepMax(200, 0)).toBe(0);
  });

  it("inverts back to a working load", () => {
    const e1rm = estimateOneRepMax(200, 5);
    expect(loadForReps(e1rm, 5)).toBeCloseTo(200, 5);
  });
});

describe("parsing prescribed work", () => {
  it("parses rep ranges and single targets", () => {
    expect(parseRepRange("8-12")).toEqual({ min: 8, max: 12 });
    expect(parseRepRange("10")).toEqual({ min: 10, max: 10 });
    expect(parseRepRange("12 each side")).toEqual({ min: 12, max: 12 });
  });

  it("returns null for timed and open-ended work", () => {
    expect(parseRepRange("30 sec")).toBeNull();
    expect(parseRepRange("AMRAP")).toBeNull();
    expect(parseRepRange(undefined)).toBeNull();
  });

  it("parses set counts, defaulting sanely", () => {
    expect(parseSetTarget("3")).toBe(3);
    expect(parseSetTarget("3-4 sets")).toBe(4);
    expect(parseSetTarget("nonsense")).toBe(3);
  });
});

describe("load increments", () => {
  it("scales the jump to the movement", () => {
    expect(loadIncrementLbs("Back Squat")).toBe(10);
    expect(loadIncrementLbs("Romanian Deadlift")).toBe(10);
    expect(loadIncrementLbs("Bench Press")).toBe(5);
    expect(loadIncrementLbs("Dumbbell Shoulder Press")).toBe(5);
    expect(loadIncrementLbs("Bicep Curl")).toBe(2.5);
    expect(loadIncrementLbs("Lateral Raise")).toBe(2.5);
  });

  it("rounds to loadable plates", () => {
    expect(roundToLoadable(183, 10)).toBe(185);
    expect(roundToLoadable(31.2, 2.5)).toBe(30);
  });
});

describe("buildExerciseProgression", () => {
  it("collapses a session to its best set and tracks the trend", () => {
    const logs = [
      log("2026-07-01", "Bench Press", 0, 185, 8),
      log("2026-07-01", "Bench Press", 1, 185, 7),
      log("2026-07-08", "Bench Press", 0, 190, 8),
      log("2026-07-15", "Bench Press", 0, 195, 9),
    ];
    const p = buildExerciseProgression(logs, "Bench Press");

    expect(p.sessions).toHaveLength(3);
    expect(p.sessions[0].topSetWeightLbs).toBe(185);
    expect(p.sessions[0].topSetReps).toBe(8);
    expect(p.trend).toBe("climbing");
    expect(p.changePct).toBeGreaterThan(0);
    expect(p.stalled).toBe(false);
  });

  it("ignores warmup sets", () => {
    const logs = [
      log("2026-07-01", "Bench Press", 0, 500, 5, undefined, "warmup"),
      log("2026-07-01", "Bench Press", 1, 185, 8),
    ];
    const p = buildExerciseProgression(logs, "Bench Press");
    expect(p.sessions).toHaveLength(1);
    expect(p.sessions[0].topSetWeightLbs).toBe(185);
  });

  it("matches exercise names case-insensitively", () => {
    const logs = [log("2026-07-01", "bench press ", 0, 185, 8)];
    expect(buildExerciseProgression(logs, "Bench Press").sessions).toHaveLength(1);
  });

  it("flags a stall after three sessions without a new best", () => {
    const logs = [
      log("2026-07-01", "Bench Press", 0, 200, 8),
      log("2026-07-08", "Bench Press", 0, 195, 8),
      log("2026-07-15", "Bench Press", 0, 195, 8),
      log("2026-07-22", "Bench Press", 0, 190, 8),
    ];
    const p = buildExerciseProgression(logs, "Bench Press");
    expect(p.sessionsSinceBest).toBe(3);
    expect(p.stalled).toBe(true);
    expect(p.trend).toBe("declining");
  });

  it("reports insufficient data when nothing usable is logged", () => {
    const p = buildExerciseProgression([log("2026-07-01", "Bench Press", 0)], "Bench Press");
    expect(p.trend).toBe("insufficient_data");
    expect(p.sessions).toHaveLength(0);
  });

  it("builds one progression per distinct exercise", () => {
    const logs = [
      log("2026-07-01", "Bench Press", 0, 185, 8),
      log("2026-07-01", "Back Squat", 0, 275, 5),
    ];
    expect(buildAllProgressions(logs)).toHaveLength(2);
  });
});

describe("prescribeNextSession", () => {
  it("asks for a baseline when there is no history", () => {
    const rx = prescribeNextSession(benchPress, undefined);
    expect(rx.action).toBe("establish_baseline");
    expect(rx.targetWeightLbs).toBeUndefined();
    expect(rx.confidence).toBe("low");
  });

  it("adds load after topping the rep range with reps to spare", () => {
    const logs = [
      log("2026-07-01", "Bench Press", 0, 185, 10, 7),
      log("2026-07-08", "Bench Press", 0, 185, 12, 8),
    ];
    const rx = prescribeNextSession(benchPress, buildExerciseProgression(logs, "Bench Press"));

    expect(rx.action).toBe("add_load");
    expect(rx.targetWeightLbs).toBe(190);
    expect(rx.targetReps).toBe(8); // resets to bottom of range
    expect(rx.rationale).toContain("190");
  });

  it("adds a rep when short of the top of the range", () => {
    const logs = [log("2026-07-08", "Bench Press", 0, 185, 9, 8)];
    const rx = prescribeNextSession(benchPress, buildExerciseProgression(logs, "Bench Press"));

    expect(rx.action).toBe("add_reps");
    expect(rx.targetWeightLbs).toBe(185);
    expect(rx.targetReps).toBe(10);
  });

  it("holds load when the last top set was a grind", () => {
    const logs = [log("2026-07-08", "Bench Press", 0, 185, 9, 10)];
    const rx = prescribeNextSession(benchPress, buildExerciseProgression(logs, "Bench Press"));

    expect(rx.action).toBe("hold");
    expect(rx.targetWeightLbs).toBe(185);
    expect(rx.rationale).toContain("RPE 10");
  });

  it("does not add load at the top of the range if the set was maximal", () => {
    const logs = [log("2026-07-08", "Bench Press", 0, 185, 12, 10)];
    const rx = prescribeNextSession(benchPress, buildExerciseProgression(logs, "Bench Press"));
    expect(rx.action).not.toBe("add_load");
  });

  it("deloads a stalled lift by 10 percent", () => {
    const logs = [
      log("2026-07-01", "Back Squat", 0, 300, 5, 8),
      log("2026-07-08", "Back Squat", 0, 290, 5, 8),
      log("2026-07-15", "Back Squat", 0, 290, 5, 9),
      log("2026-07-22", "Back Squat", 0, 285, 5, 9),
    ];
    const squat: WorkoutExercise = { name: "Back Squat", sets: "4", reps: "5-8" };
    const rx = prescribeNextSession(squat, buildExerciseProgression(logs, "Back Squat"));

    expect(rx.action).toBe("deload");
    expect(rx.targetWeightLbs).toBe(255); // 285 * 0.9 = 256.5 → nearest 5
    expect(rx.rationale).toContain("plateau");
  });

  it("suppresses load increases when recovery is low", () => {
    const logs = [log("2026-07-08", "Bench Press", 0, 185, 12, 7)];
    const progression = buildExerciseProgression(logs, "Bench Press");

    expect(prescribeNextSession(benchPress, progression).action).toBe("add_load");
    const tired = prescribeNextSession(benchPress, progression, { readinessScore: 45 });
    expect(tired.action).toBe("hold");
    expect(tired.targetWeightLbs).toBe(185);
    expect(tired.rationale).toContain("45/100");
  });

  it("scales the prescribed load by the intensity multiplier", () => {
    const logs = [log("2026-07-08", "Bench Press", 0, 200, 9, 8)];
    const progression = buildExerciseProgression(logs, "Bench Press");
    const rx = prescribeNextSession(benchPress, progression, { intensityMultiplier: 0.9 });
    expect(rx.targetWeightLbs).toBe(180);
  });

  it("skips load prescription for timed work", () => {
    const plank: WorkoutExercise = { name: "Plank", sets: "3", reps: "45 sec" };
    const rx = prescribeNextSession(plank, undefined);
    expect(rx.action).toBe("hold");
    expect(rx.targetWeightLbs).toBeUndefined();
  });

  it("marks confidence low when reps exceed the reliable e1RM window", () => {
    const highRep: WorkoutExercise = { name: "Leg Extension", sets: "3", reps: "15-20" };
    const logs = [log("2026-07-08", "Leg Extension", 0, 90, 18, 8)];
    const rx = prescribeNextSession(highRep, buildExerciseProgression(logs, "Leg Extension"));
    expect(rx.confidence).toBe("low");
  });

  it("prescribes a whole day at once", () => {
    const logs = [
      log("2026-07-08", "Bench Press", 0, 185, 12, 7),
      log("2026-07-08", "Bicep Curl", 0, 30, 12, 8),
    ];
    const day: WorkoutExercise[] = [
      benchPress,
      { name: "Bicep Curl", sets: "3", reps: "10-12" },
    ];
    const rxs = prescribeWorkoutDay(day, logs);

    expect(rxs).toHaveLength(2);
    expect(rxs[0].targetWeightLbs).toBe(190);
    expect(rxs[1].targetWeightLbs).toBe(32.5); // isolation → 2.5 lb jump
  });
});

describe("summarizeProgressions", () => {
  it("separates climbing lifts from stalled ones and surfaces recent PRs", () => {
    const logs = [
      log("2026-07-01", "Bench Press", 0, 185, 8, 8),
      log("2026-07-08", "Bench Press", 0, 195, 8, 8),
      log("2026-06-01", "Back Squat", 0, 300, 5, 8),
      log("2026-06-08", "Back Squat", 0, 290, 5, 8),
      log("2026-06-15", "Back Squat", 0, 290, 5, 8),
      log("2026-06-22", "Back Squat", 0, 285, 5, 8),
    ];
    const summary = summarizeProgressions(buildAllProgressions(logs), 14, "2026-07-10");

    expect(summary.trackedExercises).toBe(2);
    expect(summary.climbing).toContain("Bench Press");
    expect(summary.stalled).toContain("Back Squat");
    expect(summary.recentPrs.map((p) => p.exerciseName)).toEqual(["Bench Press"]);
    expect(summary.topGains[0].exerciseName).toBe("Bench Press");
  });
});

File: recomp-mobile/app/src/test/java/com/refactor/app/api/ProgressionTest.kt
package com.refactor.app.api

import com.refactor.app.api.dto.WorkoutExerciseDto
import com.refactor.app.api.dto.WorkoutSetLogDto
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/** Mirrors `src/lib/progression.test.ts` and iOS `ProgressionTests.swift` — all three must agree. */
class ProgressionTest {

    private fun log(
        date: String,
        exerciseName: String,
        setIndex: Int,
        weightLbs: Double? = null,
        reps: Int? = null,
        rpe: Double? = null,
        section: String = "main",
    ) = WorkoutSetLogDto(
        id = "$date:$exerciseName:$setIndex",
        date = date,
        planId = "plan-1",
        dayLabel = "Monday",
        section = section,
        exerciseName = exerciseName,
        globalSlot = 0,
        setIndex = setIndex,
        weightLbs = weightLbs,
        reps = reps,
        rpe = rpe,
        loggedAt = "${date}T18:00:00.000Z",
    )

    private val benchPress = WorkoutExerciseDto(name = "Bench Press", sets = "3", reps = "8-12")

    // ── e1RM ────────────────────────────────────────────

    @Test
    fun `estimateOneRepMax uses Epley without RPE`() {
        assertEquals(233.33, Progression.estimateOneRepMax(200.0, 5), 0.1)
    }

    @Test
    fun `estimateOneRepMax credits reps in reserve`() {
        assertEquals(246.67, Progression.estimateOneRepMax(200.0, 5, 8.0), 0.1)
        assertEquals(
            Progression.estimateOneRepMax(200.0, 5),
            Progression.estimateOneRepMax(200.0, 5, 10.0),
            0.001,
        )
    }

    @Test
    fun `estimateOneRepMax is zero for invalid input`() {
        assertEquals(0.0, Progression.estimateOneRepMax(0.0, 5), 0.0)
        assertEquals(0.0, Progression.estimateOneRepMax(200.0, 0), 0.0)
    }

    @Test
    fun `loadForReps inverts Epley`() {
        val e1rm = Progression.estimateOneRepMax(200.0, 5)
        assertEquals(200.0, Progression.loadForReps(e1rm, 5), 0.001)
    }

    // ── Parsing ─────────────────────────────────────────

    @Test
    fun `parseRepRange handles ranges and singles`() {
        assertEquals(Progression.RepRange(8, 12), Progression.parseRepRange("8-12"))
        assertEquals(Progression.RepRange(10, 10), Progression.parseRepRange("10"))
        assertEquals(Progression.RepRange(12, 12), Progression.parseRepRange("12 each side"))
    }

    @Test
    fun `parseRepRange returns null for timed work`() {
        assertNull(Progression.parseRepRange("30 sec"))
        assertNull(Progression.parseRepRange("AMRAP"))
        assertNull(Progression.parseRepRange(null))
    }

    @Test
    fun `parseSetTarget defaults sanely`() {
        assertEquals(3, Progression.parseSetTarget("3"))
        assertEquals(4, Progression.parseSetTarget("3-4 sets"))
        assertEquals(3, Progression.parseSetTarget("nonsense"))
    }

    // ── Load increments ─────────────────────────────────

    @Test
    fun `load increment scales to movement`() {
        assertEquals(10.0, Progression.loadIncrementLbs("Back Squat"), 0.0)
        assertEquals(10.0, Progression.loadIncrementLbs("Romanian Deadlift"), 0.0)
        assertEquals(5.0, Progression.loadIncrementLbs("Bench Press"), 0.0)
        assertEquals(5.0, Progression.loadIncrementLbs("Dumbbell Shoulder Press"), 0.0)
        assertEquals(2.5, Progression.loadIncrementLbs("Bicep Curl"), 0.0)
        assertEquals(2.5, Progression.loadIncrementLbs("Lateral Raise"), 0.0)
    }

    @Test
    fun `roundToLoadable snaps to plates`() {
        assertEquals(185.0, Progression.roundToLoadable(183.0, 10.0), 0.0)
        assertEquals(30.0, Progression.roundToLoadable(31.2, 2.5), 0.0)
    }

    // ── Trend ───────────────────────────────────────────

    @Test
    fun `buildProgression collapses sessions and tracks trend`() {
        val logs = listOf(
            log("2026-07-01", "Bench Press", 0, 185.0, 8),
            log("2026-07-01", "Bench Press", 1, 185.0, 7),
            log("2026-07-08", "Bench Press", 0, 190.0, 8),
            log("2026-07-15", "Bench Press", 0, 195.0, 9),
        )
        val p = Progression.buildExerciseProgression(logs, "Bench Press")

        assertEquals(3, p.sessions.size)
        assertEquals(185.0, p.sessions[0].topSetWeightLbs!!, 0.0)
        assertEquals(8, p.sessions[0].topSetReps)
        assertEquals(Progression.Trend.CLIMBING, p.trend)
        assertTrue(p.changePct > 0)
        assertFalse(p.stalled)
    }

    @Test
    fun `buildProgression ignores warmups`() {
        val logs = listOf(
            log("2026-07-01", "Bench Press", 0, 500.0, 5, section = "warmup"),
            log("2026-07-01", "Bench Press", 1, 185.0, 8),
        )
        val p = Progression.buildExerciseProgression(logs, "Bench Press")
        assertEquals(1, p.sessions.size)
        assertEquals(185.0, p.sessions[0].topSetWeightLbs!!, 0.0)
    }

    @Test
    fun `buildProgression matches names case-insensitively`() {
        val logs = listOf(log("2026-07-01", "bench press ", 0, 185.0, 8))
        assertEquals(1, Progression.buildExerciseProgression(logs, "Bench Press").sessions.size)
    }

    @Test
    fun `buildProgression flags stall after three sessions`() {
        val logs = listOf(
            log("2026-07-01", "Bench Press", 0, 200.0, 8),
            log("2026-07-08", "Bench Press", 0, 195.0, 8),
            log("2026-07-15", "Bench Press", 0, 195.0, 8),
            log("2026-07-22", "Bench Press", 0, 190.0, 8),
        )
        val p = Progression.buildExerciseProgression(logs, "Bench Press")
        assertEquals(3, p.sessionsSinceBest)
        assertTrue(p.stalled)
        assertEquals(Progression.Trend.DECLINING, p.trend)
    }

    @Test
    fun `buildProgression reports insufficient data`() {
        val p = Progression.buildExerciseProgression(
            listOf(log("2026-07-01", "Bench Press", 0)),
            "Bench Press",
        )
        assertEquals(Progression.Trend.INSUFFICIENT_DATA, p.trend)
        assertTrue(p.sessions.isEmpty())
    }

    @Test
    fun `buildAllProgressions returns one per exercise`() {
        val logs = listOf(
            log("2026-07-01", "Bench Press", 0, 185.0, 8),
            log("2026-07-01", "Back Squat", 0, 275.0, 5),
        )
        assertEquals(2, Progression.buildAllProgressions(logs).size)
    }

    // ── Prescription ────────────────────────────────────

    @Test
    fun `prescribes baseline without history`() {
        val rx = Progression.prescribeNextSession(benchPress, null)
        assertEquals(Progression.Action.ESTABLISH_BASELINE, rx.action)
        assertNull(rx.targetWeightLbs)
        assertEquals(Progression.Confidence.LOW, rx.confidence)
    }

    @Test
    fun `adds load after topping the rep range`() {
        val logs = listOf(
            log("2026-07-01", "Bench Press", 0, 185.0, 10, 7.0),
            log("2026-07-08", "Bench Press", 0, 185.0, 12, 8.0),
        )
        val rx = Progression.prescribeNextSession(
            benchPress,
            Progression.buildExerciseProgression(logs, "Bench Press"),
        )
        assertEquals(Progression.Action.ADD_LOAD, rx.action)
        assertEquals(190.0, rx.targetWeightLbs!!, 0.0)
        assertEquals(8, rx.targetReps)
        assertTrue(rx.rationale.contains("190"))
    }

    @Test
    fun `adds a rep when short of the range`() {
        val logs = listOf(log("2026-07-08", "Bench Press", 0, 185.0, 9, 8.0))
        val rx = Progression.prescribeNextSession(
            benchPress,
            Progression.buildExerciseProgression(logs, "Bench Press"),
        )
        assertEquals(Progression.Action.ADD_REPS, rx.action)
        assertEquals(185.0, rx.targetWeightLbs!!, 0.0)
        assertEquals(10, rx.targetReps)
    }

    @Test
    fun `holds after a grind`() {
        val logs = listOf(log("2026-07-08", "Bench Press", 0, 185.0, 9, 10.0))
        val rx = Progression.prescribeNextSession(
            benchPress,
            Progression.buildExerciseProgression(logs, "Bench Press"),
        )
        assertEquals(Progression.Action.HOLD, rx.action)
        assertEquals(185.0, rx.targetWeightLbs!!, 0.0)
        assertTrue(rx.rationale.contains("RPE 10"))
    }

    @Test
    fun `no load jump on a maximal top set`() {
        val logs = listOf(log("2026-07-08", "Bench Press", 0, 185.0, 12, 10.0))
        val rx = Progression.prescribeNextSession(
            benchPress,
            Progression.buildExerciseProgression(logs, "Bench Press"),
        )
        assertNotEquals(Progression.Action.ADD_LOAD, rx.action)
    }

    @Test
    fun `deloads a stalled lift by ten percent`() {
        val logs = listOf(
            log("2026-07-01", "Back Squat", 0, 300.0, 5, 8.0),
            log("2026-07-08", "Back Squat", 0, 290.0, 5, 8.0),
            log("2026-07-15", "Back Squat", 0, 290.0, 5, 9.0),
            log("2026-07-22", "Back Squat", 0, 285.0, 5, 9.0),
        )
        val squat = WorkoutExerciseDto(name = "Back Squat", sets = "4", reps = "5-8")
        val rx = Progression.prescribeNextSession(
            squat,
            Progression.buildExerciseProgression(logs, "Back Squat"),
        )
        assertEquals(Progression.Action.DELOAD, rx.action)
        assertEquals(255.0, rx.targetWeightLbs!!, 0.0)
        assertTrue(rx.rationale.contains("plateau"))
    }

    @Test
    fun `suppresses load increases when recovery is low`() {
        val logs = listOf(log("2026-07-08", "Bench Press", 0, 185.0, 12, 7.0))
        val progression = Progression.buildExerciseProgression(logs, "Bench Press")

        assertEquals(
            Progression.Action.ADD_LOAD,
            Progression.prescribeNextSession(benchPress, progression).action,
        )

        val tired = Progression.prescribeNextSession(
            benchPress,
            progression,
            Progression.Options(readinessScore = 45.0),
        )
        assertEquals(Progression.Action.HOLD, tired.action)
        assertEquals(185.0, tired.targetWeightLbs!!, 0.0)
        assertTrue(tired.rationale.contains("45/100"))
    }

    @Test
    fun `scales by the intensity multiplier`() {
        val logs = listOf(log("2026-07-08", "Bench Press", 0, 200.0, 9, 8.0))
        val rx = Progression.prescribeNextSession(
            benchPress,
            Progression.buildExerciseProgression(logs, "Bench Press"),
            Progression.Options(intensityMultiplier = 0.9),
        )
        assertEquals(180.0, rx.targetWeightLbs!!, 0.0)
    }

    @Test
    fun `skips load prescription for timed work`() {
        val plank = WorkoutExerciseDto(name = "Plank", sets = "3", reps = "45 sec")
        val rx = Progression.prescribeNextSession(plank, null)
        assertEquals(Progression.Action.HOLD, rx.action)
        assertNull(rx.targetWeightLbs)
    }

    @Test
    fun `marks confidence low on high reps`() {
        val highRep = WorkoutExerciseDto(name = "Leg Extension", sets = "3", reps = "15-20")
        val logs = listOf(log("2026-07-08", "Leg Extension", 0, 90.0, 18, 8.0))
        val rx = Progression.prescribeNextSession(
            highRep,
            Progression.buildExerciseProgression(logs, "Leg Extension"),
        )
        assertEquals(Progression.Confidence.LOW, rx.confidence)
    }

    @Test
    fun `prescribes a whole day keyed by normalized name`() {
        val logs = listOf(
            log("2026-07-08", "Bench Press", 0, 185.0, 12, 7.0),
            log("2026-07-08", "Bicep Curl", 0, 30.0, 12, 8.0),
        )
        val day = listOf(benchPress, WorkoutExerciseDto(name = "Bicep Curl", sets = "3", reps = "10-12"))
        val rxs = Progression.prescribeWorkoutDay(day, logs)

        assertEquals(2, rxs.size)
        assertEquals(190.0, rxs["bench press"]!!.targetWeightLbs!!, 0.0)
        assertEquals(32.5, rxs["bicep curl"]!!.targetWeightLbs!!, 0.0)
    }

    @Test
    fun `summary separates climbing from stalled`() {
        val logs = listOf(
            log("2026-07-01", "Bench Press", 0, 185.0, 8, 8.0),
            log("2026-07-08", "Bench Press", 0, 195.0, 8, 8.0),
            log("2026-06-01", "Back Squat", 0, 300.0, 5, 8.0),
            log("2026-06-08", "Back Squat", 0, 290.0, 5, 8.0),
            log("2026-06-15", "Back Squat", 0, 290.0, 5, 8.0),
            log("2026-06-22", "Back Squat", 0, 285.0, 5, 8.0),
        )
        val summary = Progression.summarize(
            Progression.buildAllProgressions(logs),
            recentDays = 14,
            today = "2026-07-10",
        )

        assertEquals(2, summary.trackedExercises)
        assertTrue(summary.climbing.contains("Bench Press"))
        assertTrue(summary.stalled.contains("Back Squat"))
        assertEquals(listOf("Bench Press"), summary.recentPrs.map { it.exerciseName })
        assertEquals("Bench Press", summary.topGains.first().exerciseName)
    }
}
